### PR TITLE
[Core][1/3] Make streaming generator public API

### PR DESCRIPTION
**Pull request by rkooo567**
## Why are these changes needed?

This PR is doing
    - [x] Introduce throw/send
    - [x] Bring up RayWaitable type
    - [x] Modify types for cancel, remote, ray.get, and ray.wait
    - [x] Update doctoring for ray.wait & ray.cancel
    - [x] Implement completed, is_next_ready, is_finished
    - [x] Allow to not specify num_returns=“streaming”

There will be total 3 PRs;
- All sort of implementation work <---- current
- Rename StreamingObjectRefGenerator and annotate it as a public API (and move it to `__init__.py`)
- Typing support

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/_private/ray_option_utils.py
Comment (rkooo567) [inline python/ray/_private/ray_option_utils.py:191]: If the function is a generator, num_returns="streaming", otherwise it will become 1 in the remote function layer

File: python/ray/_private/worker.py
Comment (rkooo567) [inline python/ray/_private/worker.py:124]: This could be contraversial.

I found a bit weird to make ObjectRef and ObjectRefGenerator a subclass of RayWaitable because

- they don't share any functionality
- they don't share any interface.

So I found it a bit more natural to just define the Union types. Any thoughts?

File: python/ray/_private/worker.py
Comment (rkooo567) [inline python/ray/_private/worker.py:2529]: Raising an exception requires lots of test changes. I will fix it in the third PR.

File: python/ray/_private/worker.py
Comment (raulchen) [reply python/ray/_private/worker.py:124]: using a union makes sense. but I think using an empty interface won't hurt either.  whichever way should be fine.

2 related questions:
1. Should `StreamingObjectRefGenerator` also be a generic type? 
2. is there any difference between defining `RayWaitable` in this way versus just `RayWaitable[R] = Union["ObjectRef[R]", StreamingObjectRefGenerator]`?

File: python/ray/_private/worker.py
Comment (raulchen) [inline python/ray/_private/worker.py:2646]: better use a generic type T instead of Any? So we can get stricter type info from type inference.

File: python/ray/_private/worker.py
Comment (raulchen) [inline python/ray/_private/worker.py:2668]: Nit, this comment gives me a feeling that ObjectRef and StreamingObjectRefGenerators are inclusive. Maybe we can phrase the comment in this way: first say the first list contains ready waitables, and the second is unready; then explain what ready/unready means for different types.

File: python/ray/_raylet.pyx
Comment (raulchen) [inline python/ray/_raylet.pyx:264]: not blocking this PR. Should we move this class to a different file? this file is pretty long. also, is there a perf improvement to define it in Cython? It seems to me that most code in this class is pure Python code.

Comment (rkooo567) [reply python/ray/_raylet.pyx:264]: I think we can definitely move out of it. I will handle it in the second PR together

File: python/ray/_private/worker.py
Comment (rkooo567) [reply python/ray/_private/worker.py:2646]: Isn't this technically wrong? My impression is T means all the Waitables will have the same type underlying data, which is not true for this APi. But I am also not very familiar with python typing.

File: python/ray/_private/worker.py
Comment (rkooo567) [reply python/ray/_private/worker.py:124]: 1. so in this case, it is the type of returned object ref? 

StreamingObjectRefGenerator[T] == `__next__(self) -> ObjectRef[T]`?

I think this sgtm.

2. So I tried defining `RayWaitable = Union["ObjectRef[R]", StreamingObjectRefGenerator]`, but it raised an exception, so I tried the current approach lol..

**Comment by rkooo567** [comment]
Btw, until @pcmoritz gives a green light, I will mark the feature as alpha. I will still proceed merging PRs needed for the proposal (and try to get it in by 2.7), so please review it! https://docs.google.com/document/d/1LzKPL7wxz5tUJJLMMZ7BlfdDlYclAzMdzolTezU06ck/edit#heading=h.ehfmi7oa003b

File: python/ray/_private/worker.py
Comment (rkooo567) [reply python/ray/_private/worker.py:2646]: addressed

File: python/ray/_private/worker.py
Comment (rkooo567) [reply python/ray/_private/worker.py:2668]: This is a nice suggestion. Can. you take a look at the revision?

File: python/ray/_private/worker.py
Comment (rkooo567) [reply python/ray/_private/worker.py:2658]: So I played with a little. Looks like having T requires us to have various overload because we can have mixture of types. I will leave it to the curernt style for now

File: python/ray/_private/worker.py
Comment (jjyao) [inline python/ray/_private/worker.py:166]: I think for the inputs, it should still be T0 or ObjectRef[T0]?

File: python/ray/_private/worker.py
Comment (jjyao) [inline python/ray/_private/worker.py:2476]: Given passing StreamingObjectRefGenerator is not allowed, we shouldn't change to RayWaitable here right?

File: python/ray/_private/worker.py
Comment (jjyao) [inline python/ray/_private/worker.py:2867]: I guess it's both waitable and cancelable lol

File: python/ray/_raylet.pyx
Comment (jjyao) [inline python/ray/_raylet.pyx:266]: Annotate with @PublicAPI

Also do we want to start with alpha or beta or stable

File: python/ray/_private/worker.py
Comment (scv119) [reply python/ray/_private/worker.py:124]: `class RayWaitable` and `Union["ObjectRef[R]", StreamingObjectRefGenerator]` are fundamentally different? looks `Union["ObjectRef[R]", StreamingObjectRefGenerator]` are more preferrable

File: python/ray/_private/worker.py
Comment (rkooo567) [reply python/ray/_private/worker.py:2658]: Let me try and see if it works with pypi

File: python/ray/_private/worker.py
Comment (rkooo567) [reply python/ray/_private/worker.py:2476]: Same answer as above. If there's a suggestion to make this work, I'd love to hear that

File: python/ray/_raylet.pyx
Comment (rkooo567) [reply python/ray/_raylet.pyx:266]: This is the scope of the second PR!

File: python/ray/_private/ray_option_utils.py
Comment (rkooo567) [reply python/ray/_private/ray_option_utils.py:190]: I think we want to hide "streaming". It will work without streaming if it is a generator. I will modify the comment when I hard deprecate the dynamic generator.

File: python/ray/_private/worker.py
Comment (rkooo567) [reply python/ray/_private/worker.py:124]: Confirmed it doesn't work. We cannot make RayWaittable a Generic type without declaring a class that inherits from Generic

File: python/ray/_private/worker.py
Comment (rkooo567) [reply python/ray/_private/worker.py:2658]: Looks like it works. I will use T

File: python/ray/_private/worker.py
Comment (rkooo567) [reply python/ray/_private/worker.py]: Addressed

File: python/ray/_private/worker.py
Comment (rkooo567) [reply python/ray/_private/worker.py]: Yep updated. I overloaded the remote method so that it can return both generator and object ref. After that it is fixed

**Comment by rkooo567** [comment]
@jjyao @scv119 

Let's get back to merging this PR.

Re; typing. 

- I remove RayWaitable and used Union instead.
- To make generator -> ObjectRefGenerator[int], it requires lots of repetitive code. I will do it in a follow up. I got it working in this commit, https://github.com/ray-project/ray/pull/38784/commits/28d535556865fe75f669a2bc50de458507b4fa9f. But I won't inclde this in this PR

File: python/ray/_private/worker.py
Comment (rkooo567) [inline python/ray/_private/worker.py:168]: This is a bug I found. When I passed object ref, mypy couldn't find the correct type, but it found ray._raylet.ObjectRef

Comment (rkooo567) [reply python/ray/_private/worker.py:168]: cc @pcmoritz

File: python/ray/types.py
Comment (rkooo567) [inline python/ray/types.py:18]: Note: Currently, Cython cannot define generic inherited classes. WE need to upgrade it to support it

File: python/ray/_private/ray_option_utils.py
Comment (jjyao) [inline python/ray/_private/ray_option_utils.py:174]:
```suggestion
        '"streaming" (for generators). "dynamic" is deprecated from Ray 2.7.',
```
?

File: python/ray/_private/worker.py
Comment (jjyao) [inline python/ray/_private/worker.py:62]: why this alias?

File: python/ray/_private/worker.py
Comment (jjyao) [inline python/ray/_private/worker.py:2688]: Does it make sense to make StreaingObjectRefGenerator generic since it can return multiple objects with different types?

File: python/ray/_raylet.pyx
Comment (jjyao) [inline python/ray/_raylet.pyx:319]: why do we need these?

File: python/ray/_private/ray_option_utils.py
Comment (jjyao) [reply python/ray/_private/ray_option_utils.py:174]: why the default value is not 1 but None, what does None mean here?

Comment (rkooo567) [reply python/ray/_private/ray_option_utils.py:174]: Let me update the comment. None basically means default. 

Default behavior is 1 for regular task, and streaming for a generator task

File: python/ray/_private/worker.py
Comment (rkooo567) [reply python/ray/_private/worker.py:62]: Because we also import ObjectRef. We currently have 2 object ref, one real one (ObjectRef in Cython), and the other just for typing (types.ObjectRef). I think we should fix this, but it is not in a scope of this PR. 

When I don't import this type, I found mypy looks for the ObjectRef in Cython, and it fails a pretty basic test when I annotate the object ref type to test_typing_good.

File: python/ray/_raylet.pyx
Comment (rkooo567) [reply python/ray/_raylet.pyx:319]: These are standard generator APIs that are not supported in Ray

File: python/ray/_private/worker.py
Comment (rkooo567) [reply python/ray/_private/worker.py:2688]: Yeah you can do StreaingObjectRefGenerator[Union[...]]

in that case

File: python/ray/_raylet.pyx
Comment (rkooo567) [reply python/ray/_raylet.pyx:319]: Also specified in https://docs.google.com/document/d/1LzKPL7wxz5tUJJLMMZ7BlfdDlYclAzMdzolTezU06ck/edit#heading=h.ehfmi7oa003b

File: python/ray/_private/worker.py
Comment (rkooo567) [reply python/ray/_private/worker.py:2688]: I think this is important decision though. Normally generator has a type annotation Generator[yield type, send type, return type], so I think it makes sense to have it in generator too. But I created a sync meeting with @pcmoritz tomorrow to discuss a bit more details.

File: python/ray/_private/ray_option_utils.py
Comment (rkooo567) [reply python/ray/_private/ray_option_utils.py:174]: updated the comment. Please take a look and lmk what you think

File: python/ray/_raylet.pyx
Comment (rkooo567) [reply python/ray/_raylet.pyx:269]: We will scope it to beta

File: python/ray/tests/typing_files/check_typing_good.py
Comment (jjyao) [inline python/ray/tests/typing_files/check_typing_good.py:63]: Why we need to disable it for now?

**Review by jjyao — APPROVED** [review]

File: python/ray/tests/typing_files/check_typing_good.py
Comment (rkooo567) [reply python/ray/tests/typing_files/check_typing_good.py:63]: It is not working now

File: python/ray/_raylet.pyx
Comment (edoakes) [inline python/ray/_raylet.pyx:4104]: better to make these kwargs

Comment (edoakes) [inline python/ray/_raylet.pyx:4104]: better to make these kwargs

Comment (rkooo567) [reply python/ray/_raylet.pyx:4104]: btw, are you talking about 

```
def __init__(self,
    method is_generator=None
):
```

something like this?

Comment (edoakes) [reply python/ray/_raylet.pyx:4108]: yes, use kwargs instead of a long list of optional positionals with comments

**Pull request merged**
